flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/StoreSinkCompactor.java

```diff
@@ -0,0 +1,135 @@
+import org.apache.flink.table.store.file.writer.RecordWriter;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/** A dedicated {@link SinkWriter} for manual triggered compaction. */
+public class StoreSinkCompactor<WriterStateT> extends StoreSinkWriterBase<WriterStateT> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(StoreSinkCompactor.class);
+
+    private final int subTaskId;
+    private final int numOfParallelInstances;
+
+    private final FileStore fileStore;
+    private final Map<BinaryRowData, Map<Integer, List<DataFileMeta>>> partitionedMeta;
+    private final Map<String, String> partitionSpec;
+
+    public StoreSinkCompactor(
+            int subTaskId,
+            int numOfParallelInstances,
+            FileStore fileStore,
+            Map<String, String> partitionSpec) {
+        this.subTaskId = subTaskId;
+        this.numOfParallelInstances = numOfParallelInstances;
+        this.fileStore = fileStore;
+        this.partitionSpec = partitionSpec;
+        this.partitionedMeta = new HashMap<>();
+    }
+
+    @Override
+    public void flush(boolean endOfInput) {
+        if (endOfInput) {
+            FileStoreScan.Plan plan =
+                    fileStore
+                            .newScan()
+                            .withPartitionFilter(
+                                    PredicateConverter.CONVERTER.fromMap(
+                                            partitionSpec, fileStore.partitionType()))
+                            .plan();
+            for (Map.Entry<BinaryRowData, Map<Integer, List<DataFileMeta>>> partEntry :
+                    plan.groupByPartFiles().entrySet()) {
+                BinaryRowData partition = partEntry.getKey();
+                for (Map.Entry<Integer, List<DataFileMeta>> bucketEntry :
+                        partEntry.getValue().entrySet()) {
+                    int bucket = bucketEntry.getKey();
+                    if (select(partition, bucket)) {
+                        if (LOG.isDebugEnabled()) {
+                            LOG.debug(
+                                    "Assign partition {}, bucket {} to subtask {}",
+                                    FileStorePathFactory.getPartitionComputer(
+                                                    fileStore.partitionType(),
+                                                    FileSystemConnectorOptions
+                                                            .PARTITION_DEFAULT_NAME
+                                                            .defaultValue())
+                                            .generatePartValues(partition),
+                                    bucket,
+                                    subTaskId);
+                        }
+                        partitionedMeta
```

Comment:
You don't need to store these meta in a class member. You can just create a Writer here.